cupyx/fallback_mode/utils.py

```diff
@@ -33,11 +42,115 @@ def _call_numpy(func, args, kwargs):
     Returns:
         Result after calling func.
     """
+    args = _convert_ndarrays(args)
 
     numpy_args, numpy_kwargs = data_transfer._get_numpy_args(args, kwargs)
 
     numpy_res = func(*numpy_args, **numpy_kwargs)
 
     cupy_res = data_transfer._get_cupy_result(numpy_res)
 
-    return cupy_res
+    return _get_fallback_ndarray(cupy_res)
+
+
+def _convert_ndarrays(args):
+    return tuple(
+        [i._array if isinstance(i, ndarray) else i for i in args])
+
+
+def _get_fallback_ndarray(res):
+
+    if isinstance(res, cp.ndarray):
+        return ndarray(res)
+
+    return res
+
+
+def make_method(name):
+    def method(self, *args, **kwargs):
+
+        args = _convert_ndarrays(args)
+
+        cupy_method = getattr(cp.ndarray, name)
+
+        res = cupy_method(self._array, *args, **kwargs)
+
+        return _get_fallback_ndarray(res)
+
+    return method
+
+
+def _create_magic_methods():
+
+    _magic_methods = [
+
+        # Comparison operators:
+        '__richcmp__',
+
+        # Truth value of an array (bool):
+        '__nonzero__',
+
+        # Unary operations:
+        '__neg__', '__pos__', '__abs__', '__invert__',
+
+        # Arithmetic:
+        '__add__', '__sub__', '__mul__', '__matmul__', '__div__',
+        '__truediv__', '__floordiv__', '__mod__', '__divmod__', '__pow__',
+        '__lshift__', '__rshift__', '__and__', '__or__', '__xor__',
+
+        # Arithmetic, in-place:
+        '__iadd__', '__isub__', '__imul__', '__idiv__', '__itruediv__',
+        '__ifloordiv__', '__imod__', '__ipow__', '__ilshift__', '__irshift__',
+        '__iand__', '__ior__', '__ixor__',
+
+        # For standard library functions:
+        '__copy__', '__deepcopy__', '__reduce__',
+
+        # Container customization:
+        '__iter__', '__len__', '__getitem__', '__setitem__',
+
+        # Conversion:
+        '__int__', '__float__', '__complex__', '__oct__', '__hex__',
+
+        # String representations:
+        '__repr__', '__str__'
+    ]
+
+    for method in _magic_methods:
+        setattr(ndarray, method, make_method(method))
+
+
+class ndarray:
+
+    def __init__(self, array):
+        self._array = array
+        self.func = None
+
+    def __getattr__(self, attr):
+
+        self.func = getattr(cp.ndarray, attr, None)
+        if not callable(self.func) and self.func is not None:
+            return getattr(self._array, attr)
+
+        if self.func is not None:
+            return self._call_cupy_ndarray
+
+        self.func = getattr(np.ndarray, attr)
+        return self._call_numpy_ndarray
+
+    def _call_cupy_ndarray(self, *args, **kwargs):
```

Comment:
Duplicates with `_call_cupy`.

Reply:
They are not exactly same. We cannot intercept `*args`, `**kwargs` in `__getattr__`. We need to return some callable to intercept arguments. Here, `_call_cupy_ndarray` is used for intercepting, which is not the case in `_call_cupy`.
Also, even if we resolve this, we need to pass `self._array` in case of `ndarray`. That would lead some sort of list and tuple operations for insertion of `self._array` in `*args`.

Reply:
How about fixing like as follows?
`return _call_cupy(self.func, *((self._array,) + args), **kwargs`

Reply:
Currently, the following error causes.
```py
>>> import numpy
>>> out = numpy.zeros(3)
>>> numpy.arange(12).reshape(3, 4).max(axis=1, out=out)
array([ 3.,  7., 11.])
>>>
>>> from cupyx.fallback_mode import numpy
>>> out = numpy.zeros(3)
>>> numpy.arange(12).reshape(3, 4).max(axis=1, out=out)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/imanishi/cupy/cupyx/fallback_mode/utils.py", line 143, in _call_cupy_ndarray
    res = self.func(self._array, *args, **kwargs)
  File "cupy/core/core.pyx", line 717, in cupy.core.core.ndarray.max
  File "cupy/core/core.pyx", line 725, in cupy.core.core.ndarray.max
  File "cupy/core/_routines_statistics.pyx", line 9, in cupy.core._routines_statistics._ndarray_max
TypeError: Argument 'out' has incorrect type (expected cupy.core.core.ndarray, got ndarray)
```
This fix resolve this error.

Reply:
This error was not related `_call_cupy` or `_call_cupy_ndarray`. I had not set the `kwargs` for conversion from `utils.ndarray` to `cupy.ndarray`. That's resolved now.
Also, I implemented `ndarray.__call` to reuse `_call_cupy` and `_call_numpy`.